src/crimson/os/seastore/onode_manager/simple-fltree/onode_delta.h

```diff
@@ -0,0 +1,70 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:nil -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <cstdint>
+
+#include "common/hobject.h"
+#include "include/buffer_fwd.h"
+
+#include "crimson/os/seastore/onode.h"
+#include "crimson/os/seastore/seastore_types.h"
+
+using crimson::os::seastore::OnodeRef;
+
+struct delta_t {
+  enum class op_t : uint8_t {
```

Comment:
> I'm not quite following why you don't just need insert_or_update(key, value), 

i think a delta is used both 

1. for stashing the pending op to a block before the mutating record is committed to disk.
1. for holding the record written to disk

the slot number will not be persisted to disk, because it will save a lookup in the block in the first use case.

regarding to "why not have insert_or_update()", i have `insert_onode` and `update_onode`. `insert_or_update` is a combination of these two ops. from the enduser of onode tree's perspective, it's more natural to have this op. but from the implementation's perspective, an `insert_or_update` will either end up as an `insert_onode` **or** `update_onode`, so it might make sense to differentiate them so we can identify potential bugs in future. for instance, we could try to insert an inode which already exist in the onode tree when replaying the records.

> remove(key, value). The contents are always ordered, right? Maybe a comment here 

i want to save the spatial overhead and the disk i/o, in case the the size of "key" plus "value" makes the size of record exceed the block size, and forces us to write two of them. and probably the overhead of memcpy when preparing the delta, 

> explaining the rationale behind the choice of mutation primitives?

sure, i will do once we have a consensus.

Reply:
Ah, so it's an optimization for applying the deltas at commit time.  With this scheme where you defer updating the buffer, your duplicate method could choose not to duplicate the bufferptr saving a copy.  However, what would happen if the same ObjectStore level transaction performs multiple mutations of different onodes within the same block?  When reading or mutating a node, you'd have to take into account any buffered deltas.

Reply:
> Ah, so it's an optimization for applying the deltas at commit time. With this scheme where you defer updating the buffer, your duplicate method could choose not to duplicate the bufferptr saving a copy. 

exactly.

> However, what would happen if the same ObjectStore level transaction performs multiple mutations of different onodes within the same block? When reading or mutating a node, you'd have to take into account any buffered deltas.

my design deliberately avoid updating the same block with multiple deltas in the same transaction.

Reply:
I'm not sure I'm following how.  I guess we could enumerate the set of onodes the transaction will touch ahead of time and then commit the final versions a single time at the end?  Is that what you mean?

Reply:
oh, Sam, i was wrong, i should support multiple deltas targeting the same block. will update this PR. a typical use case is to insert an onode in the middle of a leaf:

1. shift the onodes whose oid is greater than this one right
1. insert it

that would be two deltas.